spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/SqsMessageProducer.java

```diff
@@ -0,0 +1,160 @@
+	}
+
+	protected org.springframework.messaging.Message<String> getMessageForExecution(
+			final software.amazon.awssdk.services.sqs.model.Message message) {
+		HashMap<String, Object> additionalHeaders = new HashMap<>();
+		additionalHeaders.put(SqsMessageHeaders.SQS_LOGICAL_RESOURCE_ID, this.logicalEndpointName);
+		additionalHeaders.put(SqsMessageHeaders.RECEIVED_AT, Instant.now());
+		additionalHeaders.put(SqsMessageHeaders.QUEUE_VISIBILITY, this.queueAttributes.getVisibilityTimeout());
+		additionalHeaders.put(SqsMessageHeaders.VISIBILITY,
+				new QueueMessageVisibility(this.sqsAsyncClient, this.queueUrl, message.receiptHandle()));
+		return createMessage(message, additionalHeaders);
+	}
+
+	protected org.springframework.messaging.Message<String> createMessage(
+			software.amazon.awssdk.services.sqs.model.Message message, Map<String, Object> additionalHeaders) {
+
+		HashMap<String, Object> messageHeaders = new HashMap<>();
+		messageHeaders.put(SqsMessageHeaders.MESSAGE_ID_MESSAGE_ATTRIBUTE_NAME, message.messageId());
+		messageHeaders.put(SqsMessageHeaders.RECEIPT_HANDLE_MESSAGE_ATTRIBUTE_NAME, message.receiptHandle());
+		messageHeaders.put(SqsMessageHeaders.SOURCE_DATA_HEADER, message);
+		messageHeaders.put(MessageHeaders.ACKNOWLEDGMENT_HEADER,
+				new SqsAcknowledge(this.sqsAsyncClient, this.queueUrl, message.receiptHandle()));
+		messageHeaders.putAll(additionalHeaders);
+		messageHeaders.putAll(getAttributesAsMessageHeaders(message));
+		messageHeaders.putAll(getMessageAttributesAsMessageHeaders(message));
+		return new GenericMessage<>(message.body(), new SqsMessageHeaders(messageHeaders));
+	}
+
+	private static Map<String, Object> getMessageAttributesAsMessageHeaders(
+			software.amazon.awssdk.services.sqs.model.Message message) {
+
+		Map<String, Object> messageHeaders = new HashMap<>();
+		for (Map.Entry<MessageSystemAttributeName, String> messageAttribute : message.attributes().entrySet()) {
+			if (org.springframework.messaging.MessageHeaders.CONTENT_TYPE.equals(messageAttribute.getKey().name())) {
+				messageHeaders.put(org.springframework.messaging.MessageHeaders.CONTENT_TYPE,
+						MimeType.valueOf(messageAttribute.getValue()));
+			}
+			else if (org.springframework.messaging.MessageHeaders.ID.equals(messageAttribute.getKey().name())) {
+				messageHeaders.put(org.springframework.messaging.MessageHeaders.ID,
+						UUID.fromString(messageAttribute.getValue()));
+			}
+			else {
+				messageHeaders.put(messageAttribute.getKey().name(), messageAttribute.getValue());
+			}
+		}
+		return messageHeaders;
+	}
+
+	private static Map<String, Object> getAttributesAsMessageHeaders(
+			software.amazon.awssdk.services.sqs.model.Message message) {
+		Map<String, Object> messageHeaders = new HashMap<>();
+		for (Map.Entry<MessageSystemAttributeName, String> attributeKeyValuePair : message.attributes().entrySet()) {
+			messageHeaders.put(attributeKeyValuePair.getKey().name(), attributeKeyValuePair.getValue());
+		}
```

Comment:
I think it can be moved to make use of streams api

Reply:
Agreed. But I think there's a good chance we'll end up doing some refactoring, or perhaps rewriting, to this and other parts I've brought from the previous version, so if that's ok I prefer deferring this change for a moment.